File: shared/src/commonMain/kotlin/org/nihongo/mochi/domain/game/RecognitionGameEngine.kt
package org.nihongo.mochi.domain.game

import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.nihongo.mochi.data.ScoreManager
import org.nihongo.mochi.domain.kana.KanaToRomaji
import org.nihongo.mochi.domain.kana.KanaUtils
import org.nihongo.mochi.domain.models.AnswerButtonState
import org.nihongo.mochi.domain.models.GameStatus
import org.nihongo.mochi.domain.models.KanjiDetail
import org.nihongo.mochi.domain.models.KanjiProgress
import kotlin.random.Random

enum class QuestionDirection { NORMAL, REVERSE }

sealed class GameState {
    object Loading : GameState()
    object WaitingForAnswer : GameState()
    data class ShowingResult(val isCorrect: Boolean, val selectedAnswerIndex: Int) : GameState()
    object Finished : GameState()
}

class RecognitionGameEngine {
    var isGameInitialized = false
    
    val allKanjiDetails = mutableListOf<KanjiDetail>()
    var currentKanjiSet = mutableListOf<KanjiDetail>()
    var revisionList = mutableListOf<KanjiDetail>()
    val kanjiStatus = mutableMapOf<KanjiDetail, GameStatus>()
    val kanjiProgress = mutableMapOf<KanjiDetail, KanjiProgress>()
    var kanjiListPosition = 0
    lateinit var currentKanji: KanjiDetail
    lateinit var correctAnswer: String
    lateinit var gameMode: String
    lateinit var readingMode: String
    var currentDirection: QuestionDirection = QuestionDirection.NORMAL
    
    // UI State
    var currentAnswers = listOf<String>()
    
    private val _state = MutableStateFlow<GameState>(GameState.Loading)
    val state: StateFlow<GameState> = _state.asStateFlow()
    
    // Buttons state
    private val _buttonStates = MutableStateFlow<List<AnswerButtonState>>(List(4) { AnswerButtonState.DEFAULT })
    val buttonStates: StateFlow<List<AnswerButtonState>> = _buttonStates.asStateFlow()

    // Configuration
    var pronunciationMode: String = "Hiragana" // "Hiragana" or "Roman"
    var animationSpeed: Float = 1.0f

    fun resetState() {
        isGameInitialized = false
        allKanjiDetails.clear()
        currentKanjiSet.clear()
        revisionList.clear()
        kanjiStatus.clear()
        kanjiProgress.clear()
        kanjiListPosition = 0
        currentAnswers = emptyList()
        _state.value = GameState.Loading
        _buttonStates.value = List(4) { AnswerButtonState.DEFAULT }
    }
    
    fun startGame() {
        if (startNewSet()) {
            nextQuestion()
            _state.value = GameState.WaitingForAnswer
        } else {
            _state.value = GameState.Finished
        }
    }

    private fun startNewSet(): Boolean {
        revisionList.clear()
        kanjiStatus.clear()
        kanjiProgress.clear()

        if (kanjiListPosition >= allKanjiDetails.size) {
            return false // No more sets
        }

        val nextSet = allKanjiDetails.drop(kanjiListPosition).take(10)
        kanjiListPosition += nextSet.size

        currentKanjiSet.clear()
        currentKanjiSet.addAll(nextSet)
        revisionList.addAll(nextSet)
        currentKanjiSet.forEach {
            kanjiStatus[it] = GameStatus.NOT_ANSWERED
            kanjiProgress[it] = KanjiProgress()
        }
        return true
    }

    fun nextQuestion() {
        if (revisionList.isEmpty()) return

        currentKanji = revisionList.random()
        val progress = kanjiProgress[currentKanji]!!

        // Determine direction
        currentDirection = when {
            !progress.normalSolved && !progress.reverseSolved -> if (Random.nextDouble() < 0.5) QuestionDirection.NORMAL else QuestionDirection.REVERSE
            !progress.normalSolved -> QuestionDirection.NORMAL
            else -> QuestionDirection.REVERSE
        }

        currentAnswers = generateAnswers(currentKanji)
        _buttonStates.value = List(4) { AnswerButtonState.DEFAULT }
    }

    private fun generateAnswers(correctKanji: KanjiDetail): List<String> {
        if (currentDirection == QuestionDirection.NORMAL) {
            val correctButtonText: String
            if (gameMode == "meaning") {
                correctAnswer = correctKanji.meanings.firstOrNull() ?: ""
                correctButtonText = correctKanji.meanings.take(3).joinToString("\n")
            } else { // reading mode
                correctButtonText = getFormattedReadings(correctKanji)
                correctAnswer = correctButtonText.lines().firstOrNull() ?: ""
            }

            if (correctAnswer.isEmpty()) return listOf("", "", "", "")

            val incorrectPool = allKanjiDetails
                .asSequence()
                .filter { it.id != correctKanji.id }
                .map { detail ->
                    if (gameMode == "meaning") {
                        detail.meanings.take(3).joinToString("\n")
                    } else {
                        getFormattedReadings(detail)
                    }
                }
                .filter { it.isNotEmpty() }
                .distinct()
                .shuffled()
                .take(3)
                .toList()

            return (incorrectPool + correctButtonText).shuffled()
        } else {
            correctAnswer = correctKanji.character
            val correctButtonText = correctKanji.character

            val incorrectPool = allKanjiDetails
                .asSequence()
                .filter { it.id != correctKanji.id }
                .map { it.character }
                .distinct()
                .shuffled()
                .take(3)
                .toList()

            return (incorrectPool + correctButtonText).shuffled()
        }
    }
    
    fun getFormattedReadings(kanji: KanjiDetail): String {
        val onReadings = kanji.readings.filter { it.type == "on" }
        val kunReadings = kanji.readings.filter { it.type == "kun" }

        val selectedOn = if (readingMode == "common") {
            onReadings.sortedByDescending { it.frequency }
        } else {
            onReadings.shuffled()
        }.take(2)

        val selectedKun = if (readingMode == "common") {
            kunReadings.sortedByDescending { it.frequency }
        } else {
            kunReadings.shuffled()
        }.take(2)

        val onStrings = selectedOn.map {
            if (pronunciationMode == "Roman") KanaToRomaji.convert(it.value).uppercase() else KanaUtils.hiraganaToKatakana(it.value)
        }
        val kunStrings = selectedKun.map {
            if (pronunciationMode == "Roman") KanaToRomaji.convert(it.value) else it.value
        }

        return (onStrings + kunStrings).joinToString("\n")
    }

    suspend fun submitAnswer(selectedAnswer: String, selectedIndex: Int) {
        val isCorrect = if (currentDirection == QuestionDirection.NORMAL) {
            // For NORMAL, button text might be multi-line, check if any line matches correctAnswer
            selectedAnswer.lines().any { it.equals(correctAnswer, ignoreCase = true) }
        } else {
            selectedAnswer == correctAnswer
        }
        
        ScoreManager.saveScore(currentKanji.character, isCorrect, ScoreManager.ScoreType.RECOGNITION)
        
        // Update Game State
        val progress = kanjiProgress[currentKanji]!!
        val newButtonStates = _buttonStates.value.toMutableList()

        if (isCorrect) {
            if (currentDirection == QuestionDirection.NORMAL) progress.normalSolved = true
            else progress.reverseSolved = true

            if (progress.normalSolved && progress.reverseSolved) {
                kanjiStatus[currentKanji] = GameStatus.CORRECT
                revisionList.remove(currentKanji)
                newButtonStates[selectedIndex] = AnswerButtonState.CORRECT
            } else {
                kanjiStatus[currentKanji] = GameStatus.PARTIAL
                newButtonStates[selectedIndex] = AnswerButtonState.NEUTRAL
            }
        } else {
            kanjiStatus[currentKanji] = GameStatus.INCORRECT
            newButtonStates[selectedIndex] = AnswerButtonState.INCORRECT
        }
        
        _buttonStates.value = newButtonStates
        _state.value = GameState.ShowingResult(isCorrect, selectedIndex)
        delay((1000 * animationSpeed).toLong())

        if (revisionList.isEmpty()) {
            if (!startNewSet()) {
                _state.value = GameState.Finished
                return
            }
        }
        
        nextQuestion()
        _state.value = GameState.WaitingForAnswer
    }
}

File: flutter_app/lib/services/recognition_game_engine.dart
import 'dart:math';
import '../models/quiz_models.dart';
import '../models/dictionary.dart';
import '../repositories/score_repository.dart';
import '../utils/kana_utils.dart';

enum QuestionDirection { normal, reverse }

class KanjiProgress {
  bool normalSolved = false;
  bool reverseSolved = false;
}

class RecognitionGameEngine {
  final ScoreRepository _scoreRepo;
  final Random _random = Random();

  bool isGameInitialized = false;
  final List<DictionaryItem> allKanjiDetails = [];
  final List<DictionaryItem> currentKanjiSet = [];
  final List<DictionaryItem> revisionList = [];
  final Map<String, GameStatus> kanjiStatus = {};
  final Map<String, KanjiProgress> kanjiProgress = {};
  
  int kanjiListPosition = 0;
  DictionaryItem? currentKanji;
  
  int correctAnswersInSession = 0;
  int totalAnswersInSession = 0;

  late String correctAnswer;
  late String gameMode;
  late String readingMode;
  QuestionDirection currentDirection = QuestionDirection.normal;
  
  List<String> currentAnswers = [];
  GameState state = GameState.loading;
  List<AnswerButtonState> buttonStates = List.generate(4, (_) => AnswerButtonState.defaultState);
  int errorCount = 0;
  double animationSpeed = 1.0;

  String pronunciationMode = "Hiragana"; // "Hiragana" or "Roman"
  bool isProcessingAnswer = false;

  RecognitionGameEngine(this._scoreRepo);

  void resetState() {
    isGameInitialized = false;
    allKanjiDetails.clear();
    currentKanjiSet.clear();
    revisionList.clear();
    kanjiStatus.clear();
    kanjiProgress.clear();
    kanjiListPosition = 0;
    currentAnswers = [];
    currentKanji = null;
    isProcessingAnswer = false;
    errorCount = 0;
    state = GameState.loading;
    buttonStates = List.generate(4, (_) => AnswerButtonState.defaultState);
    correctAnswersInSession = 0;
    totalAnswersInSession = 0;
  }

  List<GameStatus> get currentSetStatus => 
    currentKanjiSet.map((k) => kanjiStatus[k.character] ?? GameStatus.notAnswered).toList();

  void startGame() {
    if (allKanjiDetails.isEmpty) {
      state = GameState.finished;
      return;
    }

    kanjiStatus.clear();
    kanjiProgress.clear();
    for (var item in allKanjiDetails) {
      kanjiStatus[item.character] = GameStatus.notAnswered;
      kanjiProgress[item.character] = KanjiProgress();
    }

    if (startNewSet()) {
      nextQuestion();
      state = GameState.waitingForAnswer;
    } else {
      state = GameState.finished;
    }
  }

  bool startNewSet() {
    revisionList.clear();
    if (kanjiListPosition >= allKanjiDetails.length) return false;

    final nextSet = allKanjiDetails.skip(kanjiListPosition).take(10).toList();
    kanjiListPosition += nextSet.length;

    currentKanjiSet.clear();
    currentKanjiSet.addAll(nextSet);
    revisionList.addAll(nextSet);
    return true;
  }

  void nextQuestion() {
    if (revisionList.isEmpty) {
      if (!startNewSet()) {
        state = GameState.finished;
        return;
      }
    }

    currentKanji = revisionList[_random.nextInt(revisionList.length)];
    final progress = kanjiProgress[currentKanji!.character]!;

    if (!progress.normalSolved && !progress.reverseSolved) {
      currentDirection = _random.nextBool() ? QuestionDirection.normal : QuestionDirection.reverse;
    } else if (!progress.normalSolved) {
      currentDirection = QuestionDirection.normal;
    } else {
      currentDirection = QuestionDirection.reverse;
    }

    currentAnswers = _generateAnswers(currentKanji!);
    buttonStates = List.generate(4, (_) => AnswerButtonState.defaultState);
  }

  List<String> _generateAnswers(DictionaryItem correctKanji) {
    if (currentDirection == QuestionDirection.normal) {
      String correctButtonText;
      if (gameMode == "meaning") {
        correctAnswer = correctKanji.meanings.isNotEmpty ? correctKanji.meanings.first : "";
        correctButtonText = correctKanji.meanings.take(3).join("\n");
      } else {
        correctButtonText = getFormattedReadings(correctKanji);
        correctAnswer = correctButtonText.split("\n").first;
      }

      if (correctAnswer.isEmpty) return ["", "", "", ""];

      final incorrectPool = allKanjiDetails
          .where((k) => k.id != correctKanji.id)
          .map((detail) {
            if (gameMode == "meaning") {
              return detail.meanings.take(3).join("\n");
            } else {
              return getFormattedReadings(detail);
            }
          })
          .where((s) => s.isNotEmpty)
          .toSet()
          .toList()
        ..shuffle(_random);

      final result = incorrectPool.take(3).toList()..add(correctButtonText);
      return result..shuffle(_random);
    } else {
      correctAnswer = correctKanji.character;
      final correctButtonText = correctKanji.character;

      final incorrectPool = allKanjiDetails
          .where((k) => k.id != correctKanji.id)
          .map((k) => k.character)
          .toSet()
          .toList()
        ..shuffle(_random);

      final result = incorrectPool.take(3).toList()..add(correctButtonText);
      return result..shuffle(_random);
    }
  }

  String getFormattedReadings(DictionaryItem kanji) {
    final onReadings = kanji.readings.where((r) => r.type == 'on').toList();
    final kunReadings = kanji.readings.where((r) => r.type == 'kun').toList();

    if (readingMode == "common") {
      // In a real app, we might have frequency data for readings too.
    } else {
      onReadings.shuffle(_random);
      kunReadings.shuffle(_random);
    }

    final onStrings = onReadings.take(2).map((r) => KanaUtils.hiraganaToKatakana(r.text)).toList();
    final kunStrings = kunReadings.take(2).map((r) => r.text).toList();

    return (onStrings + kunStrings).join("\n");
  }

  Future<bool> submitAnswer(int selectedIndex) async {
    final kanji = currentKanji;
    if (kanji == null || isProcessingAnswer || state == GameState.finished) return false;
    isProcessingAnswer = true;

    final selectedAnswer = currentAnswers[selectedIndex];
    final isCorrect = currentDirection == QuestionDirection.normal
        ? selectedAnswer.split("\n").any((line) => line.trim().toLowerCase() == correctAnswer.trim().toLowerCase())
        : selectedAnswer == correctAnswer;

    totalAnswersInSession++;
    if (isCorrect) {
      correctAnswersInSession++;
    } else {
      errorCount++;
    }

    await _scoreRepo.saveScore(
      key: kanji.character,
      wasCorrect: isCorrect,
      type: gameMode == "meaning" ? ScoreType.recognition : ScoreType.reading,
    );

    final progress = kanjiProgress[kanji.character]!;
    if (isCorrect) {
      if (currentDirection == QuestionDirection.normal) {
        progress.normalSolved = true;
      } else {
        progress.reverseSolved = true;
      }

      if (progress.normalSolved && progress.reverseSolved) {
        kanjiStatus[kanji.character] = GameStatus.correct;
        revisionList.remove(kanji);
        buttonStates[selectedIndex] = AnswerButtonState.correct;
      } else {
        kanjiStatus[kanji.character] = GameStatus.partial;
        buttonStates[selectedIndex] = AnswerButtonState.neutral;
      }
    } else {
      kanjiStatus[kanji.character] = GameStatus.incorrect;
      buttonStates[selectedIndex] = AnswerButtonState.incorrect;
      
      final correctIdx = currentAnswers.indexWhere((ans) => currentDirection == QuestionDirection.normal 
        ? ans.split("\n").any((line) => line.trim().toLowerCase() == correctAnswer.trim().toLowerCase())
        : ans == correctAnswer
      );
      if (correctIdx != -1) buttonStates[correctIdx] = AnswerButtonState.correct;
    }

    state = GameState.showingResult;
    return isCorrect;
  }

  void acknowledgeResult() {
    isProcessingAnswer = false;
    if (revisionList.isEmpty) {
      if (!startNewSet()) {
        state = GameState.finished;
        return;
      }
    }
    nextQuestion();
    state = GameState.waitingForAnswer;
  }
}
